src/scene/materials/shader-graph-builder.js

```diff
@@ -0,0 +1,128 @@
+
+Object.assign(ShaderGraphBuilder.prototype, {
+    /**
+     * @private
+     * @function
+     * @name pc.ShaderGraphBuilder#addParam
+     * @description adds a parameter input to graph
+     * @param {string} type - type of parameter
+     * @param {string} name - name of parameter
+     * @param {any} value - optional value of parameter
+     * @returns {any} returns the created input port
+     */
+    addParam: function (type, name, value) {
+        var ioPort = this._graph.addInput(type, name, value);
+        return ioPort;
+    },
+    /**
+     * @private
+     * @function
+     * @name pc.ShaderGraphBuilder#addNode
+     * @description creates and adds a core node to the shader graph and connects up inputs
+     * @param {string} name - name of parameter
+     * @param {object} [arguments] - first argument is name string, then 0-N output node ports or params to connect to new core node inputs - number and port type of arguments must match.
+     * @returns {number} returns the created node id
+     * @example
+     * var node0 = pc.shadergraph.addNode('uv0');
+     * var node1 = pc.shadergraph.addNode('texSample', param0, node0 );
+     * var node2 = pc.shadergraph.addNode('splitVec4', node1 );
+     * var node3 = pc.shadergraph.addNode('mul3', param1, node42 );
+     */
+    addNode: function (name) {
+        var args = arguments;
+        var coreNode = this._nodes[name];
+        var sgIndex = this._graph.addSubGraph(coreNode);
+
+        var sgInputs = coreNode.graphData.ioPorts.filter(function (ioPort) {
+            return ioPort.name.startsWith('IN_');
+        });
+
+        if (sgInputs.length === (args.length - 1)) {
+            for (var argIndex = 0; argIndex < (args.length - 1); argIndex++) {
+                var arg = args[argIndex + 1];
+                if (typeof(arg) === 'number') {
+                    // ret port of a node
+                    var argNodeIndex = arg;
+                    this._graph.connect(argNodeIndex, 'OUT_ret', sgIndex, sgInputs[argIndex].name);
+                } else if (arg.type) {
+                    // ioPort (ioPort)
+                    this._graph.connect(-1, arg.name, sgIndex, sgInputs[argIndex].name);
+                } else {
+                    // specific port of a node
+                    this._graph.connect(arg.node, 'OUT_' + arg.port, sgIndex, sgInputs[argIndex].name);
+                }
+            }
+        } else {
+            console.warn("arguments do not match core node");
+        }
+        return sgIndex;
+    },
+    /**
+     * @private
+     * @function
+     * @name pc.ShaderGraphBuilder#addOutput
+     * @description connect a node port to graph material output
+     * @param {any} arg - node port to output
+     * @param {string} type - type of output
+     * @param {string} name - name of output
+     * @example
+     * // valid outputs are:
+     * pc.shadergraph.output(arg, 'vec3', 'vertOff');

```

Comment:
I think valid outputs should be stored in a global array as data.
Something like
`var validOutputs = [ { name: "dAlpha", type: "float"}, ... ];`
This would give us:
- easier validation, and remove the need to pass in type
- would allow us to implement function to do this instead code like this being used in few places in other files:
`        if (rootShaderGraph.getIoPortByName('OUT_dAlpha') || rootShaderGraph.getIoPortByName('OUT_dNormalMap') || rootShaderGraph.getIoPortByName('OUT_dGlossiness') || rootShaderGraph.getIoPortByName('OUT_dSpecularity') || rootShaderGraph.getIoPortByName('OUT_dAlbedo') || rootShaderGraph.getIoPortByName('OUT_fragOut') || rootShaderGraph.getIoPortByName('OUT_dEmission')) {
`

Reply:
yes - when the API is made public, a public enum should be added - probably the same with the core nodes - as mentioned before, used privately with the editor there will be no mismatch - until custom outputs are added to the editor (after MVP)

Reply:
After more thought, the shader graph builder and the shader graph shouldn't the outputs - this should be done at a higher level by the system that binds the outputs - which for the MVP will the standard node material - in the future, the shader graph will be used by other systems like the particle materials and post processes. Therefore, I altered the examples given in the JSDoc